### ISSUE TITLE
Assertion failure in _endBackgroundTask when running in iOS 8.3 simulator

### ISSUE THREAD

**Issue by al-s** [swisspol/GCDWebServer#170]
Steps to reproduce.

- Create a simple app with GCDWebServer. Start the server with the option `GCDWebServerOption_AutomaticallySuspendInBackground` set to `NO`.
```objc
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    self.webServer = [[GCDWebServer alloc] init];
    [self.webServer addDefaultHandlerForMethod:@"GET" requestClass:[GCDWebServerRequest class] processBlock:^GCDWebServerResponse *(GCDWebServerRequest *request) {
        return [GCDWebServerDataResponse responseWithHTML:@"<html><body><p>Hello World!</p></body></html>"];
    }];
	NSDictionary *options = @{
		GCDWebServerOption_Port: @(8080),
		GCDWebServerOption_AutomaticallySuspendInBackground: @(NO)
	};
	NSError *error;
	if ([self.webServer startWithOptions:options error:&error])
		NSLog(@"Visit %@ in your web browser", self.webServer.serverURL);
	else
		NSLog(@"%@", error);
    return YES;
}
```
- Start the application in iOS 8.3 simulator.
- Send the application to background mode.
- Open the web server page using Safari under iOS or under MacOS while the app is in background.
- Application crashes on assertion inside `[GCDWebServer _endBackgroundTask]`:
```objc
// Always called on main thread
- (void)_endBackgroundTask {
  GWS_DCHECK([NSThread isMainThread]); // <-- here
  if (_backgroundTask != UIBackgroundTaskInvalid) {
…
```
which is very strange because documentation states that handler for `[UIApplication beginBackgroundTaskWithExpirationHandler]` "is called synchronously on the main thread".

In iOS 8.2 simulator this code works without problems. Looks like a bug in iOS 8.3: under certain conditions it can call the handler not on the main thread.

It seems that solution is to call `beginBackgroundTaskWithExpirationHandler` when app goes to background if `GCDWebServerOption_AutomaticallySuspendInBackground` is set to `NO`. Currently it is called for individual connections, but in this situation it also should be called for the entire server because it continues to listen on the socket.

**Comment by swisspol** [comment]
> In iOS 8.2 simulator this code works without problems. Looks like a bug in iOS 8.3: under certain conditions it can call the handler not on the main thread.

Can you copy-paste all the stack traces of the process (`bt all` in lldb)?

> which is very strange because documentation states that handler for [UIApplication beginBackgroundTaskWithExpirationHandler] "is called synchronously on the main thread".

Can you file a bug to https://bugreport.apple.com/ and provide the number? Seems like this is going to break a number of apps.

> It seems that solution is to call beginBackgroundTaskWithExpirationHandler when app goes to background

Why would that be the solution?

> Currently it is called for individual connections, but in this situation it also should be called for the entire server because it continues to listen on the socket.

That's not needed as the server will close the socket once the last connection closes.

**Comment by al-s** [comment]
> Can you copy-paste all the stack traces of the process (bt all in lldb)?

```
  thread #1: tid = 0x15ae4, 0x000000011183352e libsystem_kernel.dylib`mach_msg_trap + 10, queue = 'com.apple.main-thread'
    frame #0: 0x000000011183352e libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x000000011183269f libsystem_kernel.dylib`mach_msg + 55
    frame #2: 0x000000010fad2624 CoreFoundation`__CFRunLoopServiceMachPort + 212
    frame #3: 0x000000010fad1ad7 CoreFoundation`__CFRunLoopRun + 1287
    frame #4: 0x000000010fad1366 CoreFoundation`CFRunLoopRunSpecific + 470
    frame #5: 0x0000000112ceda3e GraphicsServices`GSEventRunModal + 161
    frame #6: 0x00000001100f6900 UIKit`UIApplicationMain + 1282
    frame #7: 0x000000010eeb192f TestWebServer`main(argc=1, argv=0x00007fff50d6f338) + 111 at main.m:14
    frame #8: 0x0000000111519145 libdyld.dylib`start + 1
    frame #9: 0x0000000111519145 libdyld.dylib`start + 1

  thread #2: tid = 0x15b08, 0x000000011183922e libsystem_kernel.dylib`kevent64 + 10, queue = 'com.apple.libdispatch-manager'
    frame #0: 0x000000011183922e libsystem_kernel.dylib`kevent64 + 10
    frame #1: 0x00000001114d376c libdispatch.dylib`_dispatch_mgr_invoke + 247
    frame #2: 0x00000001114d3511 libdispatch.dylib`_dispatch_mgr_thread + 54

  thread #8: tid = 0x15be2, 0x000000010ee95110 TestWebServer`+[GCDWebServerDataResponse(self=0x000000010eeb8e80, _cmd=0x00007ff54a53aa80, html=0x000000010eeb8e80) responseWithHTML:] at GCDWebServerDataResponse.m:88, queue = 'com.apple.root.default-qos'
    frame #0: 0x000000010ee95110 TestWebServer`+[GCDWebServerDataResponse(self=0x000000010eeb8e80, _cmd=0x00007ff54a53aa80, html=0x000000010eeb8e80) responseWithHTML:] at GCDWebServerDataResponse.m:88
    frame #1: 0x000000010eeae4de TestWebServer`__57-[AppDelegate application:didFinishLaunchingWithOptions:]_block_invoke(.block_descriptor=0x000000010eeb8e80, request=0x00007ff54a53aa80) + 78 at AppDelegate.m:26
    frame #2: 0x000000010ee9ce77 TestWebServer`__79-[GCDWebServer(.block_descriptor=0x00007ff54a6197e0, request=0x00007ff54a53aa80, completionBlock=0x000000011b5648c0) addDefaultHandlerForMethod:requestClass:processBlock:]_block_invoke + 119 at GCDWebServer.m:797
    frame #3: 0x000000010eead549 TestWebServer`-[GCDWebServerConnection(self=0x00007ff54a421a30, _cmd=0x000000010eeb4525, request=0x00007ff54a53aa80, completion=0x000000011b5648c0) processRequest:completion:] + 425 at GCDWebServerConnection.m:762
    frame #4: 0x000000010eea939f TestWebServer`-[GCDWebServerConnection _startProcessingRequest](self=0x00007ff54a421a30, _cmd=0x000000010eeb45b2) + 271 at GCDWebServerConnection.m:384
    frame #5: 0x000000010eeaba50 TestWebServer`__45-[GCDWebServerConnection _readRequestHeaders]_block_invoke(.block_descriptor=0x00007ff54a513ab0, extraData=0x00007ff54a65ca30) + 3424 at GCDWebServerConnection.m:586
    frame #6: 0x000000010eea695d TestWebServer`__65-[GCDWebServerConnection(.block_descriptor=<unavailable>, success=true) _readHeaders:withCompletionBlock:]_block_invoke + 701 at GCDWebServerConnection.m:133
    frame #7: 0x000000010eea6151 TestWebServer`__69-[GCDWebServerConnection(.block_descriptor=0x00007ff54a5f10d0, buffer=0x00007ff54a420470, error=0) _readData:withLength:completionBlock:]_block_invoke + 385 at GCDWebServerConnection.m:103
    frame #8: 0x00000001114deaaa libdispatch.dylib`__dispatch_read_block_invoke_252 + 39
    frame #9: 0x00000001114c5186 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #10: 0x00000001114e4614 libdispatch.dylib`_dispatch_client_callout + 8
    frame #11: 0x00000001114ce552 libdispatch.dylib`_dispatch_root_queue_drain + 1768
    frame #12: 0x00000001114cfb17 libdispatch.dylib`_dispatch_worker_thread3 + 111
    frame #13: 0x00000001118696cb libsystem_pthread.dylib`_pthread_wqthread + 729
    frame #14: 0x00000001118674a1 libsystem_pthread.dylib`start_wqthread + 13

  thread #9: tid = 0x15be7, 0x0000000111838946 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #0: 0x0000000111838946 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x0000000111869757 libsystem_pthread.dylib`_pthread_wqthread + 869
    frame #2: 0x00000001118674a1 libsystem_pthread.dylib`start_wqthread + 13

* thread #7: tid = 0x15be8, 0x0000000111838282 libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.root.user-initiated-qos.overcommit', stop reason = signal SIGABRT
    frame #0: 0x0000000111838282 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x000000011186b4c3 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00000001115c719a libsystem_sim_c.dylib`abort + 129
  * frame #3: 0x000000010ee981a3 TestWebServer`-[GCDWebServer _endBackgroundTask](self=0x00007ff54a618de0, _cmd=0x000000010eeb329e) + 51 at GCDWebServer.m:280
    frame #4: 0x000000010ee97c50 TestWebServer`__36-[GCDWebServer _startBackgroundTask]_block_invoke(.block_descriptor=<unavailable>) + 112 at GCDWebServer.m:229
    frame #5: 0x0000000110114361 UIKit`___addBackgroundTask_block_invoke_2 + 64
    frame #6: 0x0000000117ed29d5 AssertionServices`__60-[BKSProcessAssertion _clientQueue_notifyAssertionAcquired:]_block_invoke + 20
    frame #7: 0x00000001114c5186 libdispatch.dylib`_dispatch_call_block_and_release + 12
    frame #8: 0x00000001114e4614 libdispatch.dylib`_dispatch_client_callout + 8
    frame #9: 0x00000001114ce552 libdispatch.dylib`_dispatch_root_queue_drain + 1768
    frame #10: 0x00000001114cfb17 libdispatch.dylib`_dispatch_worker_thread3 + 111
    frame #11: 0x00000001118696cb libsystem_pthread.dylib`_pthread_wqthread + 729
    frame #12: 0x00000001118674a1 libsystem_pthread.dylib`start_wqthread + 13
```

> Why would that be the solution?

Because it just works :) I tried to call `beginBackgroundTaskWithExpirationHandler` in my code before the app goes to background, and application had about 3 minutes to finish its work, expiration handler was called on the main thread. Without this "global" background task, `beginBackgroundTaskWithExpirationHandler` is called only by `[GCDWebServer _startBackgroundTask]`. In this case its expiration handler is called almost immediately and not on the main thread. Another strange thing is that `[UIApplication backgroundTimeRemaining]` inside this handler returns 1.797693134862316e+308.

> That's not needed as the server will close the socket once the last connection closes.

Perhaps this happens only when `GCDWebServerOption_AutomaticallySuspendInBackground` is set to `YES`? Otherwise, if listening socket is closed, how can I connect to the server from the browser when application is already in background and I have not created any connections yet?

**Comment by swisspol** [comment]
> Because it just works :)

That's a work around not a fix though.

> Otherwise, if listening socket is closed, how can I connect to the server from the browser when application is already in background and I have not created any connections yet?

You shouldn't do that because that's very unreliable considering iOS can freeze your app at any time. Worse if you don't close the socket when the app is frozen, clients will just hang on connection.

Everything is explained here:
https://github.com/swisspol/GCDWebServer/blob/master/README.md#gcdwebserver--background-mode-for-ios-apps

**Comment by swisspol** [comment]
This bug should be fixed now (a3606d6027c038771c0622dfb8a70cd95e0a30c3), thanks for reporting it!
